src/cbor/CBOREncoder.cpp

```diff
@@ -45,24 +46,68 @@ CborError CBOREncoder::encode(PropertyContainer & property_container, uint8_t *
    */
   CborError error = CborNoError;
   int num_encoded_properties = 0;
-  std::for_each(property_container.begin(),
-                property_container.end(),
-                [lightPayload, &arrayEncoder, &error, &num_encoded_properties](Property * p)
-                {
-                  if (p->shouldBeUpdated() && p->isReadableByCloud())
-                  {
-                    error = p->append(&arrayEncoder, lightPayload);
-                    if(error == CborNoError)
-                      num_encoded_properties++;
-                    else
-                      return;
-                  }
-                });
-  if ((CborNoError != error) && 
-      (CborErrorOutOfMemory != error))
+  int num_checked_properties = 0;
+  static int encoded_properties_message_limit = CBOR_ENCODER_NO_PROPERTIES_LIMIT;
+
+  if(current_property_index >= property_container.size())
+    current_property_index = 0;
+
+  PropertyContainer::iterator iter = property_container.begin();
+  std::advance(iter, current_property_index);
+
+  for(; iter != property_container.end(); iter++)
+  {
+    Property * p = * iter;
+    bool maximum_number_of_properties_reached = (num_encoded_properties >= encoded_properties_message_limit) && (encoded_properties_message_limit != -1);
+    bool cbor_encoder_error = (error != CborNoError);
+
+    if (maximum_number_of_properties_reached || cbor_encoder_error)
+      break;
+
+    if (p->shouldBeUpdated() && p->isReadableByCloud())
+    {
+      error = p->append(&arrayEncoder, lightPayload);
+      if(error == CborNoError)
+        num_encoded_properties++;
+    }
+    num_checked_properties++;
+  }
+
+  if ((CborNoError != error) && (CborErrorOutOfMemory != error))
+  {
+    /* Trim the number of properties to be included in the next message to avoid multivalue property split */
+    encoded_properties_message_limit = num_encoded_properties;
     return error;
+  }
+
+  error = cbor_encoder_close_container(&encoder, &arrayEncoder);
+  if (CborNoError != error)
+  {
+    /* Trim the number of properties to be included in the next message to avoid error closing container */
+    encoded_properties_message_limit = num_encoded_properties - 1;
```

Comment:
The change feels like it uses `encoded_properties_message_limit` as both counter AND status variable. Can you think of a way to change the PR to `encoded_properties_message_limit` only as a counter (and possibly rename it) and use a boolean variable to cover the cases `CBOR_ENCODER_NO_PROPERTIES_LIMIT` and `-1`. If `num_encoded_properties - 1` is also a valid state than this should be encoded as a [FSM](https://en.wikipedia.org/wiki/Finite-state_machine) making the states and state transitions explicit. Right now states and state transitionings are hidden in the business logic and really hard to see.